File: source/shared-json/src/test/java/com/chronos/mobile/source/sharedjson/DefaultTimetableShareCodecTest.kt
package com.chronos.mobile.source.sharedjson

import com.chronos.mobile.core.model.AcademicConfig
import com.chronos.mobile.core.model.Course
import com.chronos.mobile.core.model.Timetable
import com.chronos.mobile.core.model.TimetableImportMetadata
import com.chronos.mobile.core.model.TimetableImportSource
import com.chronos.mobile.core.model.TimetableViewPrefs
import com.chronos.mobile.domain.AcademicCalendarService
import com.chronos.mobile.domain.TimeProvider
import com.chronos.mobile.domain.result.AppResult
import java.time.LocalDate
import java.time.LocalTime
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class DefaultTimetableShareCodecTest {
    private val codec = DefaultTimetableShareCodec(
        academicCalendarService = AcademicCalendarService(),
        timeProvider = FixedTimeProvider,
    )

    @Test
    fun `encode and decode preserve key timetable semantics`() {
        val timetable = sampleTimetable(importSource = TimetableImportSource.SHARED_JSON)

        val encoded = codec.encode(timetable)
        val payload = (codec.decode((encoded as AppResult.Success).value) as AppResult.Success).value
        val imported = (codec.toTimetable(payload) as AppResult.Success).value

        assertEquals(timetable.name, payload.yearTerm)
        assertEquals(TimetableImportSource.SHARED_JSON.name, payload.importSource)
        assertEquals("2026-03-02", payload.termStartDate)
        assertEquals(TimetableImportSource.SHARED_JSON, imported.importMetadata.source)
        assertEquals("2026-03-02", imported.academicConfig.termStartDate)
        assertEquals(true, imported.viewPrefs.showSaturday)
    }

    @Test
    fun `online import export omits term start date`() {
        val timetable = sampleTimetable(importSource = TimetableImportSource.ONLINE_EDU)
        val payload = (codec.decode((codec.encode(timetable) as AppResult.Success).value) as AppResult.Success).value

        assertEquals(TimetableImportSource.ONLINE_EDU.name, payload.importSource)
        assertEquals(null, payload.termStartDate)
    }

    @Test
    fun `decode rejects payload without courses`() {
        val result = codec.toTimetable(com.chronos.mobile.core.model.OnlineSchedulePayload())
        assertTrue(result is AppResult.Failure)
    }

    private fun sampleTimetable(importSource: TimetableImportSource): Timetable = Timetable(
        id = "t1",
        name = "2025-2026学年第2学期",
        courses = listOf(
            Course(
                id = "c1",
                name = "编译原理",
                teacher = "张老师",
                location = "B201",
                dayOfWeek = 6,
                startPeriod = 1,
                endPeriod = 2,
                color = "#EADDFF",
                weeks = listOf(1, 2, 3),
            ),
        ),
        createdAt = 1L,
        updatedAt = 1L,
        academicConfig = AcademicConfig(
            termStartDate = "2026-03-02",
            startWeek = 1,
            endWeek = 20,
        ),
        importMetadata = TimetableImportMetadata(source = importSource),
        viewPrefs = TimetableViewPrefs(showSaturday = true, showSunday = false),
    )
}

private object FixedTimeProvider : TimeProvider {
    override fun today(): LocalDate = LocalDate.parse("2026-03-18")
    override fun currentTime(): LocalTime = LocalTime.parse("09:00")
    override fun currentTimeMillis(): Long = 100L
}

File: src/lib/parsers/shared-json/default-timetable-share-codec.test.ts
import { describe, expect, it } from 'vite-plus/test';
import { emptyOnlineSchedulePayload } from '$lib/models/online-schedule';
import { TimetableImportSource } from '$lib/models/timetable';
import { DefaultTimetableShareCodec } from './default-timetable-share-codec';
import type { TimeProvider } from '$lib/domain/services/time-provider';

const fixedTimeProvider: TimeProvider = {
	today: () => '2026-03-18',
	currentTime: () => '09:00',
	currentTimeMillis: () => 100
};

describe('DefaultTimetableShareCodec', () => {
	const codec = new DefaultTimetableShareCodec(undefined, fixedTimeProvider);

	it('toTimetable preserves key timetable semantics and course data', () => {
		const payload = {
			...emptyOnlineSchedulePayload(),
			yearTerm: '2025-2026学年第2学期',
			weekNum: '3',
			nowMonth: '3',
			importSource: TimetableImportSource.ONLINE_EDU,
			termStartDate: '2026-03-02',
			yearTermList: ['2025-2026学年第2学期'],
			weekList: ['1', '2', '3'],
			weekDayList: [{ weekDay: '六', weekDate: '03/21', today: false }],
			eventList: [
				{
					weekNum: '3',
					weekDay: '6',
					weekList: ['1', '2', '3'],
					weekCover: '1-3周',
					sessionList: ['1', '2'],
					sessionStart: '1',
					sessionLast: '2',
					eventName: '编译原理',
					address: 'B201',
					memberName: '张老师',
					remark: '带教材第 3 版',
					duplicateGroupType: 'none',
					duplicateGroup: 0,
					eventType: 'course',
					eventID: 'c1'
				}
			]
		};

		const imported = codec.toTimetable(payload);
		expect(imported.ok).toBe(true);
		if (!imported.ok) return;

		expect(imported.value.name).toBe('2025-2026学年第2学期');
		expect(imported.value.importMetadata.source).toBe(TimetableImportSource.ONLINE_EDU);
		expect(imported.value.academicConfig.termStartDate).toBe('2026-03-02');
		expect(imported.value.viewPrefs.showSaturday).toBe(true);
		expect(imported.value.courses[0]?.name).toBe('编译原理');
		expect(imported.value.courses[0]?.remark).toBe('带教材第 3 版');
	});

	it('toTimetable infers term start date when not explicitly provided', () => {
		const payload = {
			...emptyOnlineSchedulePayload(),
			yearTerm: '2025-2026-2',
			weekNum: '1',
			nowMonth: '3',
			importSource: TimetableImportSource.ONLINE_EDU,
			termStartDate: null,
			yearTermList: ['2025-2026-2'],
			weekList: ['1'],
			weekDayList: [{ weekDay: '一', weekDate: '03/02', today: true }],
			eventList: [
				{
					weekNum: '1',
					weekDay: '1',
					weekList: ['1'],
					weekCover: '1周',
					sessionList: ['1'],
					sessionStart: '1',
					sessionLast: '1',
					eventName: '编译原理',
					address: 'B201',
					memberName: '张老师',
					remark: '',
					duplicateGroupType: '',
					duplicateGroup: 0,
					eventType: 'course',
					eventID: '1'
				}
			]
		};

		const imported = codec.toTimetable(payload);
		expect(imported.ok).toBe(true);
		if (!imported.ok) return;

		expect(imported.value.academicConfig.termStartDate).toBe('2026-03-02');
	});

	it('toTimetable rejects payload without courses', () => {
		const result = codec.toTimetable({
			yearTerm: '',
			weekNum: '',
			nowMonth: '',
			importSource: '',
			termStartDate: null,
			yearTermList: [],
			weekList: [],
			weekDayList: [],
			eventList: []
		});
		expect(result.ok).toBe(false);
	});

	it('toTimetable applies campus period times from fetch context', () => {
		const payload = {
			...emptyOnlineSchedulePayload(),
			yearTerm: '2025-2026-2',
			importSource: TimetableImportSource.ONLINE_EDU,
			weekList: ['1'],
			eventList: [
				{
					weekNum: '1',
					weekDay: '一',
					weekList: ['1'],
					weekCover: '',
					sessionList: ['1'],
					sessionStart: '1',
					sessionLast: '1',
					eventName: '编译原理',
					address: 'B201',
					memberName: '张老师',
					remark: '',
					duplicateGroupType: '',
					duplicateGroup: 0,
					eventType: 'course',
					eventID: '1'
				}
			]
		};

		const imported = codec.toTimetable(payload, {
			campusId: 'huaxi',
			campusPeriodTimes: {
				liangjiang: [{ index: 1, startTime: '08:30', endTime: '09:15' }],
				huaxi: [{ index: 1, startTime: '08:00', endTime: '08:45' }]
			}
		});
		expect(imported.ok).toBe(true);
		if (!imported.ok) return;

		expect(imported.value.importMetadata.campusId).toBe('huaxi');
		expect(imported.value.academicConfig.periodTimes).toEqual([
			{ index: 1, startTime: '08:00', endTime: '08:45' }
		]);
	});
});
